rfcs/97-add-maximum-file-size-for-storage-write.md

```diff
@@ -0,0 +1,145 @@
+---
+author: JinnyYi <github.com/JinnyYi>
+status: draft
+updated_at: 2021-06-09
+---
+
+# GSP-97: Add Maximum File Size for Storage Write
+
+## Background
+
+Depending on the size of the data you are uploading, services offer different options, like upload a file in a single operation, in parts, or through append operation.
+
+We use `Write` to handle single write operation. And the size is not unlimited.
+
+- S3 alike storage services have 5GB limit for a single PUT operation.
+- azblob limited to 5000 MiB(<https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob>).
+
+If we upload a file with size out of limit in a single operation, we will get an error like `Request Entity Too Large` in azblob. Same with `copy`,`fetch`, etc.
+
+We should figure out the error before sending request to reduce the consumption of network resources.
+
+These restrictions belong to service. We have `StorageMeta` to carry storage related information:
+
+```go
+type StorageMeta struct {
+	location string
+	Name     string
+	WorkDir  string
+	
+	// bit used as a bitmap for object value, 0 means not set, 1 means set 
+	bit uint64
+	m   map[string]interface{}
+}
+```
+
+But the current `StorageMeta` cannot solve the above problem:
+
+- The fields are finite. For now, `m` is unused, and it's not convenient to hold the extra unknown metadata.
+- If we want to use storage metadata internally we have to call `metadata` repeatedly.
+
+## Proposal
+
+So I propose to add maximum size restriction as service related metadata for storage operations.
+
+We will add `service-metadata` with type `any` into [info_storage_meta.toml].
+
+```toml
+[service-metadata]
+type = "any"
+description = "serviceMetadata stores service related metadata"
+```
+
+So that `serviceMetadata` with type `interface{}` will be generated and added into `StorageMeta` to hold the restrictions and possible settings in the future.
+
+```go
+type StorageMeta struct {
+	location        string
+	Name            string
+	WorkDir         string
+	serviceMetadata interface{}
+	
+	// bit used as a bitmap for object value, 0 means not set, 1 means set 
+	bit uint64
+}
+```
+
+To assign `serviceMetadata`, [go-storage] should generate the `StorageMetadata` structure through the field `[infos.storage.meta.<storage meta name>]` like the `ObjectMetadata`. `set/get` functions are the same.
+
+```toml
+[infos.storage.meta.<storage meta name>]
+type = "<type>"
+description = "<description>"
+```
+
+Take add maximum file size for storage write operation as an example.
+
+We should add the following in `service.toml`:
+
+```toml
+[infos.storage.meta.maxinum-write-size]
```

Comment:
`size-maximum` LGTM, this meta should also affect other operations like copy/move.

BTW, `size-maximum` should be a global storage meta instead of service meta?

Reply:
Maybe we need serval pairs to distinguish the limitations of different API? For ali oss, the maximum size of `PutObject` is 5GB but `CopyObject` is 1GB, for example.

Reply:
Ooops, I don't know about this.

So we will need `write-size-maximum` / `copy-size-maximum` and so on.